explorer/interpreter/resolve_unformed.cpp

```diff
@@ -0,0 +1,234 @@
+            ResolveUnformed(opt_exp.arguments().front(), flow_facts,
+                            /*set_formed=*/true));
+      } else {
+        for (Nonnull<Expression*> operand : opt_exp.arguments()) {
+          CARBON_RETURN_IF_ERROR(ResolveUnformed(operand, flow_facts));
+        }
+      }
+      break;
+    }
+    case ExpressionKind::DotSelfExpression:
+    case ExpressionKind::IntLiteral:
+    case ExpressionKind::BoolLiteral:
+    case ExpressionKind::BoolTypeLiteral:
+    case ExpressionKind::IntTypeLiteral:
+    case ExpressionKind::StringLiteral:
+    case ExpressionKind::StringTypeLiteral:
+    case ExpressionKind::TypeTypeLiteral:
+    case ExpressionKind::ContinuationTypeLiteral:
+    case ExpressionKind::ValueLiteral:
+    case ExpressionKind::IndexExpression:
+    case ExpressionKind::SimpleMemberAccessExpression:
+    case ExpressionKind::CompoundMemberAccessExpression:
+    case ExpressionKind::IfExpression:
+    case ExpressionKind::WhereExpression:
+    case ExpressionKind::StructLiteral:
+    case ExpressionKind::StructTypeLiteral:
+    case ExpressionKind::IntrinsicExpression:
+    case ExpressionKind::UnimplementedExpression:
+    case ExpressionKind::FunctionTypeLiteral:
+    case ExpressionKind::ArrayTypeLiteral:
+    case ExpressionKind::InstantiateImpl:
+      break;
+  }
+  return Success();
+}
+
+static auto ResolveUnformed(
+    Nonnull<Pattern*> pattern,
+    std::unordered_map<ValueNodeView, FlowFact>& flow_facts,
+    const bool has_init) -> ErrorOr<Success> {
+  switch (pattern->kind()) {
+    case PatternKind::BindingPattern:
+      flow_facts.insert(
+          {ValueNodeView(&cast<BindingPattern>(*pattern)), {has_init}});
+      break;
+    case PatternKind::TuplePattern:
+      for (Nonnull<Pattern*> field : cast<TuplePattern>(*pattern).fields()) {
+        CARBON_RETURN_IF_ERROR(ResolveUnformed(field, flow_facts, has_init));
+      }
+      break;
+    case PatternKind::GenericBinding:
+    case PatternKind::AlternativePattern:
+    case PatternKind::ExpressionPattern:
+    case PatternKind::AutoPattern:
+    case PatternKind::VarPattern:
+    case PatternKind::AddrPattern:
+      // do nothing
+      break;
+  }
+  return Success();
+}
+
+static auto ResolveUnformed(
+    Nonnull<Statement*> statement,
+    std::unordered_map<ValueNodeView, FlowFact>& flow_facts)
+    -> ErrorOr<Success> {
+  switch (statement->kind()) {
+    case StatementKind::Block: {
+      auto& block = cast<Block>(*statement);
+      for (auto* block_statement : block.statements()) {
+        CARBON_RETURN_IF_ERROR(ResolveUnformed(block_statement, flow_facts));
+      }
+      break;
+    }
+    case StatementKind::VariableDefinition: {
+      auto& def = cast<VariableDefinition>(*statement);
+      CARBON_RETURN_IF_ERROR(ResolveUnformed(&def.pattern(), flow_facts,
+                                             /*has_init=*/def.has_init()));
+      break;
+    }
+    case StatementKind::ReturnVar:
+      break;
```

Comment:
Is this deliberately doing nothing? Should it check the variable?

Is `return var` tested for uninitialized state?

Reply:
This is one of the flows that haven't been implemented yet. I plan to do them on another PR. Will add an comment to specify the unimplemented state.